### DISCUSSION TITLE
1.0.0-2

### DISCUSSION THREAD

**Discussion by JoernT** [Jinntec/Fore#96] [General]
# 1.0.0-2

## Performance Updates

* fixed some issues with mainGraph construction
* tracking changed modelItems
* subgraphing for recalculation so that only nodes are re-computed that are dependant of changed nodes

## Fixes

* Submission response dispatches `submit-error` for response status > 400Gr
* updated to latest dep_graph.js
* `targetref` wrongly used instance object instead of firstChildElement
* template finding in controls
* regression in evaluation of control nodeset
* missing constraint property on ModelItem
* inscopeContext evaluation returning first entry from an array now
* accessing defaultContext is prevented during model.init

## Additions

* new Fore Element reference describing attributes, event and link to demo usages
* Actions may attach listeners on window
* `xpath-default-namespace` can be set individually on each `fx-instance` to allow mixing of namespaces
* id references in repeats are resolved now so that individual elements can be referenced within a repeat template
* still experimental lazy refresh behind a flag attribute `refresh-on-view` that will refresh only reference that are currently in the viewport
* changed `fx-alert` to plain vanilla component
* new `fx-inspector` to log all live instance data to a section at the bottom of the viewport 
* basic `context()` function - still needs further testcases
* prettyfied XML output with `log()` function
* allowing to insert a response node into a target node with submission
* `fx-instance` may have type `html` now - for outputting somewhere in the page with `fx-output`
* subgraph demo with visualization of dependency graph
* refactor: renamed formElement to foreElement 
* preventing updateModel when no binds are present avoiding construction of dependency graph and update cycle altogether

<hr /><em>This discussion was created from the release <a href='https://github.com/Jinntec/Fore/releases/tag/v1.0.0-2'>1.0.0-2</a>.</em>

**Comment by josefwetzel** [comment]
This new version 1.0.0-2 is great! Now Fore also works on my machine! 

Version 1.0.0-1 didn't work. It didn't show any of the demo pages. (Just the "Fore" startpage and the "element reference" page.)
I work on a Macbook with BigSur 11.6.2 and eXist-db 6.0.0 and use the Chrome browser updated to the latest version.
I did the preparation steps as described in the "readme" on this fore github page (npm i and npm run start) in version 1.0.0-1 and 1.0.0-2). 

But anyway, now with 1.0.0-2 it works as expected and I am thrilled. I work easily with XForms and like the declarative approach very much. It seems easy for me to replace e.g. xf:instance with fx-instance and so on. 

I will now try to get my project to run under Fore. It was started in betterForm (about 10 years ago), in the end it reached about 2000 lines of XForms code and about 20 instances. I then migrated it to XSLTForms (about 6 months ago, because eXist-db stopped to support betterForm) and now I might "come back" to Fore.
I began to study some of the examples and the "element reference" page and I am confident that I can migrate my XForms application with reasonable effort.

Will come back to this Discussion soon again ;-).
 
Josef Wetzel

**Reply by JoernT** [reply]
Josef,

nice to hear. Please let me know if you run into issues or questions.

Joern
